opm/simulators/flow/partitionCells.cpp

```diff
@@ -357,6 +363,122 @@ partitionCellsZoltan(const int                                      num_domains,
+
+            for (const auto& cell : inputIx) {
+                const auto* cellVal = &rawInput[3*cell + 0];
+                groupedCells.push_back(cellVal[1]); // Cartesian index
+                groupedCells.push_back(cellVal[2]); // Block ID
+            }
+
+            const auto& start = cellGroup.startPointers();
+            std::copy(start.begin(), start.end(), startPtr.begin());
+        }
+
+        comm.broadcast(startPtr.data(), comm.size() + 1, 0);
+
+        // We're scattering two ints per cell
+        for (auto& startIx : startPtr) {
+            startIx *= 2;
+        }
+
+        auto sendLength = std::vector<int>(comm.size());
+        std::adjacent_difference(startPtr.begin() + 1,
+                                 startPtr.end(),
+                                 sendLength.begin());
+
+        auto perProcessInput = std::vector<int>(sendLength[comm.rank()], 0);
+        comm.scatterv(groupedCells.data(),
+                      sendLength.data(),
+                      startPtr.data(),
+                      perProcessInput.data(),
+                      static_cast<int>(perProcessInput.size()),
+                      0);
+
+        return perProcessInput;
+    }
+
+    template <class GridView, class Element>
+    std::vector<int>
+    scatterPartition(const std::vector<int>&                        rawInput,
+                     const GridView&                                grid_view,
+                     const Opm::ZoltanPartitioningControl<Element>& zoltan_ctrl)
+    {
+        auto g2l = std::unordered_map<int,int>{};
+        {
+            auto localCell = 0;
+
+            for (const auto& cell : elements(grid_view, Dune::Partitions::interior)) {
+                const auto globIx = zoltan_ctrl.local_to_global(zoltan_ctrl.index(cell));
+                g2l.insert_or_assign(globIx, localCell++);
+            }
+        }
+
+        const auto myPartitionInput = groupInputByProcessID(rawInput, grid_view.comm());
+
+        auto partition = std::vector<int>(g2l.size(), -1);
+        for (auto i = 0*myPartitionInput.size(); i < myPartitionInput.size(); i += 2) {
+            auto cellPos = g2l.find(myPartitionInput[i + 0]);
+            if (cellPos != g2l.end()) {
+                partition[cellPos->second] = myPartitionInput[i + 1];
+            }
+        }
+
+        return partition;
+    }
+} // Anonymous namespace
+
+#if HAVE_MPI
+
+namespace {
+    template <class GridView, class Element>
+    std::pair<std::vector<int>, int>
+    partitionCellsFromFileMPI(const std::filesystem::path&                   fileName,
+                              const GridView&                                grid_view,
+                              const Opm::ZoltanPartitioningControl<Element>& zoltan_ctrl)
+    {
+        const auto input = (grid_view.comm().rank() == 0)
+            ? integerVectorFromFile(fileName)
+            : std::vector<int>{};
+
+        return Opm::util::compressAndCountPartitionIDs
```

Comment:
There should be some more testing for invalid input. I think the `input.size()` should be checked in this scope, and then all individual MPI ranks' sizes verified against the size of the scattered data in scatterPartition().